### DISCUSSION TITLE
Export & import for prompts, copy between projects

### DISCUSSION THREAD

**Discussion by zoltan-fedor** [langfuse/langfuse#1696] [Ideas]
### Describe the feature or potential improvement

We have more and more prompts and it would be nice if we could export them (a whole list of them, not just an individual prompt) and then in another Langfuse instance or just in another project of the same Langfuse instance import them.

### Additional information

This would help to migrate a set of prompts between multiple environments.

Prompts imported should overwrite prompts with the same name.

**Comment by marcklingen** [comment]
Coincidentally chatted today with a team having the same idea/issue when using prompt management across multiple environments. Currently they rely on managing prompts in one project while using tracing in multiple projects.

We’re currently considering adding an organization level to simplify permissions when having multiple projects. Copying records (prompts, dataset items) between projects within an org seems to be a very natural/useful extension here.

**Reply by pannaf** [reply]
+1 here- this would be super useful

**Comment by zoltan-fedor** [comment]
> Copying records (prompts, dataset items) between projects within an org seems to be a very natural/useful extension here.
Yes, I very much agree.
Either directly copying prompts (preferable more than 1 at a time) between projects or just a simple export/import functionality for prompts would mostly address this need.

**Comment by GMartin-dev** [comment]
Agree for prompts and datasets, there is a need  for easily export / import at least.

**Comment by meissnereric** [comment]
The core usecase we have for this is promoting a new prompt from dev -> prod environment. Worried about mis-matched changes during deployments otherwise without having to write up our own sort of version-switching code for dev/prod management.

**Reply by marcklingen** [reply]
we'll add dev/prod labels for prompt versions this week which should make it easy to tag prompts for different envs within the same langfuse project. Organizations will be shipped soon, then enabling between project copying of prompts will be easier.

**Comment by touhi99** [comment]
would love to have this feature. we tried bunch of prompts and all traces in our dev branch. and when we wished to move to prod, we didnt want to lose old prompt versioning and some stats.

**Comment by j10sanders** [comment]
@marcklingen Is there currently an ability to export/copy evaluators between projects? We've created a bunch in our non-prod project and want to move them over. I don't see a way to do that?

**Reply by marcklingen** [reply]
Currently, this is not supported. Could you create a separate thread for this?

**Comment by YaEvan** [comment]
Due to the difference in costs between cloud servers and development servers, we've also encountered the need to export old data from langfuse cloud servers to development servers. We hope there can be convenient import and export features. Also, thanks for providing the langfuse tool.

**Comment by skabbit** [comment]
In our case we are looking for feature to move trace from one project to another (from dev/test to prod). Simply changing the project_id in trace table in DB is not enough, and I worry about damaging DB this way.
Is there any other option to do that?

**Reply by marcklingen** [reply]
This is not supported unfortunately as the projectId are part of the ordering key within Langfuse. You can add the traces to a new project after reading them via the sdks/api.

**Reply by skabbit** [reply]
Thanks for answer!
"reading them via the sdks/api" - do you mean that Python API support loading from one project a trace to some Python object representing the structure and then saving this object to another project?
Or you mean make it manually by reproducing all necessary steps of trace?

**Reply by marcklingen** [reply]
> "reading them via the sdks/api" - do you mean that Python API support loading from one project a trace to some Python object representing the structure and then saving this object to another project?

You can use `fetch_trace` and `fetch_observation`, then employ the low-level Python SDK to add them to another project. However, this process can be tedious since it is not a common use case. If you believe the documentation could be improved or if you could provide an example script that you used, that would be very helpful.

Notebook on the fetch_ methods that could be extended: https://langfuse.com/guides/cookbook/example_query_data_via_sdk

**Comment by jakelucasnyc** [comment]
+1 for this idea. This would be super useful not only for transferring prompts to other langfuse instances (I manage multiple separate self-hosted instances), but also being able to use the UI to back up a list of prompts to a flat file for safe keeping. Is this on the roadmap, @marcklingen?

**Comment by pi-infected** [comment]
Hi, 

I would like to be able to export all my prompts. Just for the fact that it feel safer to have my code elsewhere than to rely on a database which can have data corruption, SSD spontaneous combustion or HDD clicks of death.

I love to be able to manage my prompt on a UI and to be able to change them easily in prod, but prompts in the code have the advantage to be safe thanks to github. 

Prompts in the database are much more harder to keep safe.

**Comment by david-AH** [comment]
+1

Ability to share or sanely manage the migration of prompts across projects is a critical missing feature for us as well.  Ideally we could create, version and pull+use prompts from one project, but have all of the usage traces in a separate project.

**Comment by SuperKenVery** [comment]
I met this problem too. So I wrote this script: https://gist.github.com/SuperKenVery/1d96bc79181b3e1778dda85dc4e879c6

I've also wrote instructions on how to use it there. Hope this helps!

**Comment by danilotpnta** [comment]
Considering switching to LangWatch for this feature:

https://langwatch.ai/docs/prompt-management/cli

Hopefully this get support, as I need the prompts for easy access in my repository for my devs to have a quick look.
